src/test/java/org/primeframework/mvc/util/QueryStringBuilderTest.java

```diff
@@ -73,6 +73,48 @@ public void testBuilder() {
     test("http://acme.com/", b -> b.withSegment("bar").beginQuery().with("foo", "bar").beginFragment().with("bing", "baz"), "http://acme.com/bar?foo=bar#bing=baz");
     test("http://acme.com/", b -> b.withSegment("bar").withSegment("baz").beginQuery().with("foo", "bar").beginFragment().with("bing", "baz"), "http://acme.com/bar/baz?foo=bar#bing=baz");
 
+    // URL with query term, adding another term
+    test("http://acme.com?test=data", b -> b.with("foo", "bar"), "http://acme.com?test=data&foo=bar");
+    test("http://acme.com?test=data", b -> b.beginQuery().with("foo", "bar"), "http://acme.com?test=data&foo=bar");
+
+    // URL with fragment, adding more fragment terms
+    test("http://acme.com#frag", b -> b.with("foo", "bar"), "http://acme.com#frag&foo=bar");
+    test("http://acme.com#frag", b -> b.with("foo", "bar").with("bing", "baz"), "http://acme.com#frag&foo=bar&bing=baz");
+    test("http://acme.com#frag", b -> b.beginFragment().with("foo", "bar").with("bing", "baz"), "http://acme.com#frag&foo=bar&bing=baz");
+
+    // Should not include query since no terms were added before beginFragment
+    test("http://acme.com", b -> b.beginQuery().beginFragment().with("foo", "bar"), "http://acme.com#foo=bar");
+
+    // URLs contain trailing control characters
+    test("http://acme.com?", b -> b.with("foo", "bar"), "http://acme.com?foo=bar");
+    test("http://acme.com#test&", b -> b.beginFragment().with("foo", "bar"), "http://acme.com#test&foo=bar");
+    test("http://acme.com?test=data&", b -> b.beginQuery().with("foo", "bar"), "http://acme.com?test=data&foo=bar");
+
+    // Begin query and fragement but add no terms
+    test("http://acme.com", QueryStringBuilder::beginQuery, "http://acme.com");
+    test("http://acme.com", QueryStringBuilder::beginFragment, "http://acme.com");
+
+    try {
+      test("http://acme.com?test=data", b -> b.withSegment("bar").with("foo", "bar"), "http://acme.com/bar?test=data&foo=bar");
+      fail("Expected this to fail");
+    } catch (Exception e) {
+      assertEquals(e.getMessage(), "You cannot add a URL segment after you have appended a ? to the end of the URL");
+    }
+
+    try {
+      test("http://acme.com#frag", b -> b.withSegment("bar").with("foo", "bar"), "http://acme.com/bar#frag");
+      fail("Expected this to fail");
+    } catch (Exception e) {
+      assertEquals(e.getMessage(), "You cannot add a URL segment after you have appended a # to the end of the URL");
+    }
```

Comment:
Ideally you'd add another method to this test harness to do this work. This will be cleaner and make it easier to add additional failure case tests in the future. 

For example:

```
// You cannot add a segment after a fragment. 
testInvalid("http://acme.com#frag", b -> b.withSegment("bar").with("foo", "bar"),
            "You cannot add a URL segment after you have appended a # to the end of the URL");
```

```                   
private void testInvalid(String uri, Consumer<QueryStringBuilder> consumer, String expectedMessage) {
  try {
    QueryStringBuilder builder = QueryStringBuilder.builder(uri);
    consumer.accept(builder);
    fail("Expected this to fail.");
  } catch (Exception e) {
    assertEquals(e.getMessage(), expectedMessage);
  }
}                   
```